Use DOCKER_HOST to find the correct address for exposed container ports

OwenTuz opened the pull request:
Fixes #670 

This has only been tested on my Mac but I believe the test should now pass when using `docker-machine` (or other remote Docker hosts) as well as native Docker on both Linux and Windows.

brikis98 reviewed (commented):
Ah, this is a good idea, thanks!

I think it makes sense to take the code you wrote and expose it in a method called something like `docker.GetHost`, as I'm sure we'll want to use this functionality in the future! You can then use that new method in this test, and all other Docker-related tests in the future.

OwenTuz commented:
Makes sense to me, I'll refactor and push that change.

OwenTuz commented on modules/docker/host_test.go:
```diff
@@ -0,0 +1,66 @@
+		t.Run(fmt.Sprintf("GetDockerHost: %s", test.Input), func(t *testing.T) {
+			// GetHost() uses the DOCKER_HOST environment variable, so we need to set
+			// it to our test input and then reset it afterwards for other tests
+			defer os.Setenv("DOCKER_HOST", os.Getenv("DOCKER_HOST"))
```
perhaps it's worth clarifying that setting and resetting the environment variable means we can't safely use `t.Parallel()` here?

I believe it's safe to do this provided we don't call `Parallel()` but somebody with more experience here may well be able to correct me on that.

brikis98 replied:
Alternative idea I often use in this case:

1. Create a `getDockerHostFromEnv(env []string) string` method that takes in a list of environment variables (each one in the `key=value` format).
1. Update `GetDockerHost` to call `getDockerHost(os.Environ())` so it's using the real env vars.
1. Update the automated tests to pass in mock values.

brikis98 commented on modules/docker/host_test.go:
```diff
@@ -0,0 +1,66 @@
+
+			host := GetDockerHost()
+
+			if host != test.Expected {
+				t.Fatalf("Error: expected %s, got %s", test.Expected, host)
+			}
```
NIT: you can use `assert.Equal(t, test.Expected, host)`

brikis98 commented on modules/docker/stop_test.go:
```diff
@@ -20,7 +20,7 @@ func TestStop(t *testing.T) {
 	// choosing a unique port since 80 may not fly well on test machines
 	port := "13030"
-	testURL := "http://localhost:" + port
+	testURL := "http://" + GetDockerHost() + ":" + port
```
NIT: I know the code originally was written with concat, but it's more idiomatic in Go to use:

```suggestion
	testURL := fmt.Sprintf("http://%s:%d", GetDockerHost(), port)
```

OwenTuz commented:
Thanks for the comments - I've incorporated these and also added some handling for the corner case where `DOCKER_HOST` is not just empty but unset. If we're in that state then something else is probably broken, but we can at least handle our end safely enough by returning a default value of "localhost".

I've also made a complete mess with my merge commits, sorry: left the commits for now to show my working but happy to squash this to a single commit and re-push if you would like.

brikis98 commented on modules/docker/host.go:
```diff
@@ -0,0 +1,40 @@
+
+	for _, item := range env {
+		envVar := strings.Split(item, "=")
+		if envVar[0] == "DOCKER_HOST" {
```
Need to check that `len(envVar) == 2` as well.

brikis98 commented on modules/docker/host.go:
```diff
@@ -0,0 +1,40 @@
+
+	switch dockerUrl[0] {
+	case "tcp", "ssh", "fd":
+		return strings.TrimPrefix(dockerUrl[1], "//")
```
Need to check `len(dockerUrl) > 1`

brikis98 replied:
I know you're checking that `len(dockerUrl)` above is not 0, but that still doesn't mean there are at least two items here, so we need that check somewhere...

OwenTuz replied:
This is checked on line 24 if/when we apply `strings.Split` - but could be clearer, I'll refactor it to a single check. Thanks!

brikis98 commented on modules/docker/host_test.go:
```diff
@@ -0,0 +1,76 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetDockerHostFromEnv(t *testing.T) {
```
Can add `t.Parallel()` here and in all the `t.Run` below.

brikis98 replied:
Still need a `t.Parallel()` here, at the top of the main test method.

OwenTuz commented:
D'oh. Thanks for your patience with this one! I've made those changes, and got rid of the merge commits - this could still use being squashed on merge, but to hopefully make review a little easier.

brikis98 commented on modules/docker/host.go:
```diff
@@ -0,0 +1,44 @@
+
+	for _, item := range env {
+		envVar := strings.Split(item, "=")
+		if envVar[0] == "DOCKER_HOST" && len(envVar) == 2 {
```
NIT: I happen to know `strings.Split` returns a non-empty array, but since we're checking length anyway, it's safer to do that length check first.

```suggestion
		if len(envVar) == 2 && envVar[0] == "DOCKER_HOST" {
```

brikis98 commented:
OK, great, thank you! I'll kick off tests now.

brikis98 reviewed (approved):
There were some intermittent test failures unrelated to this PR, but on re-run, they passed, so merging now!

brikis98 commented:
https://github.com/gruntwork-io/terratest/releases/tag/v0.30.11
